lib/Support/Log.cpp

```diff
@@ -0,0 +1,84 @@
+/**
+ * Copyright (c) 2017-present, Facebook, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "glow/Support/Log.h"
+
+namespace glow {
+
+llvm::cl::opt<bool>
+    dumpCompilationLogOpt("dump-compilation-log",
+                          llvm::cl::desc("Dump compilation log"));
+
+void LogContext::addLogContent(llvm::StringRef logContent) {
+  logContents_.push_back(logContent);
+}
+
+void LogContext::addLogScope(llvm::StringRef scopeName) {
+  logScopes_.push_back(scopeName);
+  currentFullScope_ = "";
+  for (auto s : logScopes_) {
+    currentFullScope_ += "->";
+    currentFullScope_ += s;
+  }
+}
+
+void LogContext::popBackLogScope() {
```

Comment:
Similarly, why do you need to rebuild `currentFullScope_` each time? Can't you discard the last `logScopes_.back().size()` characters of `currentFullScope_` plus a few more for the `-->`?